internal/database/connections/live/migration_test.go

```diff
@@ -46,37 +50,111 @@ func testMigrations(t *testing.T, name string, schema *schemas.Schema) {
+					// Run down to the root "squashed commits" migration. We don't go
+					// any farther than that because it would require a fresh database,
+					// and that doesn't adequately test upgrade idempotency.
+					TargetVersions: []int{schema.Definitions.Root().ID},
+				},
 			},
-		},
-	}
+		}
+		if err := migrationRunner.Run(ctx, options); err != nil {
+			t.Fatalf("failed to perform downgrade: %s", err)
+		}
+	})
 
-	//
-	// Run migrations up -> down -> up
+	all := schema.Definitions.All()
 
-	if err := migrationRunner.Run(ctx, upOptions); err != nil {
-		t.Fatalf("failed to perform initial upgrade: %s", err)
-	}
-	if err := migrationRunner.Run(ctx, downOptions); err != nil {
-		t.Fatalf("failed to perform downgrade: %s", err)
+	t.Run("idempotent up", func(t *testing.T) {
+		for i, definition := range all {
+			options := runner.Options{
+				Operations: []runner.MigrationOperation{
+					{
+						SchemaName:     name,
+						Type:           runner.MigrationOperationTypeTargetedUp,
+						TargetVersions: []int{definition.ID},
+					},
+				},
+			}
+			if err := migrationRunner.Run(ctx, options); err != nil {
+				t.Fatalf("failed to perform upgrade to version %d: %s", definition.ID, err)
+			}
+
+			if i == 0 {
+				// Skip root migration
+				continue
+			}
+
+			if _, err := db.Exec(definition.UpQuery.Query(sqlf.PostgresBindVar)); err != nil {
+				t.Errorf("migration %d is not idempotent%s: %s", definition.ID, formatHint(err), err)
+			}
+		}
+	})
+
+	t.Run("idempotent down", func(t *testing.T) {
+		// Skip root migration
+		for i := len(all) - 1; i > 0; i-- {
+			definition := all[i]
+
+			options := runner.Options{
+				Operations: []runner.MigrationOperation{
+					{
+						SchemaName:     name,
+						Type:           runner.MigrationOperationTypeTargetedDown,
+						TargetVersions: definition.Parents,
+					},
+				},
+			}
+			if err := migrationRunner.Run(ctx, options); err != nil {
+				t.Fatalf("failed to perform downgrade to versions %v: %s", definition.Parents, err)
+			}
+			if _, err := db.Exec(definition.DownQuery.Query(sqlf.PostgresBindVar)); err != nil {
+				t.Errorf("migration %d is not idempotent%s: %s", definition.ID, formatHint(err), err)
+			}
+		}
+	})
+}
+
+func formatHint(err error) string {
+	var pgErr *pgconn.PgError
+	if !errors.As(err, &pgErr) {
+		return ""
 	}
-	if err := migrationRunner.Run(ctx, upOptions); err != nil {
-		t.Fatalf("failed to perform subsequent upgrade: %s", err)
+
+	switch pgErr.Code {
+	case
+		// `undefined_*` error codes
+		"42703", "42883", "42P01",
+		"42P02", "42704":
+
+		return ` (hint: use "IF EXISTS" in deletion statements)`
+
+	case
+		// `duplicate_*` error codes
+		"42701", "42P03", "42P04",
+		"42723", "42P05", "42P06",
+		"42P07", "42712", "42710":
+
+		return ` (hint: use "IF NOT EXISTS"/"CREATE OR REPLACE" in creation statements (tables, views, functions), or drop existing objects prior to creating them (indexes, constraints, triggers))`
```

Comment:
Should we really encourage dropping an index to be idempotent? 😅

Reply:
Oops that's not even in the right list.

```suggestion
		return ` (hint: use "IF NOT EXISTS"/"CREATE OR REPLACE" in creation statements (tables, indexes, views, functions), or drop existing objects prior to creating them (user-defined types, constraints, triggers))`

```